File: app/src/main/java/es/santirivera/surveilfall/base/presenter/BasePresenter.java
package es.santirivera.surveilfall.base.presenter;

import android.content.Context;
import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import butterknife.ButterKnife;
import es.santirivera.surveilfall.base.activity.BaseActivity;
import es.santirivera.surveilfall.base.interfaces.BaseNavigation;
import es.santirivera.surveilfall.base.view.BaseView;
import es.santirivera.surveilfall.domain.use_cases.base.UseCaseHandler;
import es.santirivera.surveilfall.domain.use_cases.providers.UseCaseProvider;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import javax.inject.Inject;


@SuppressWarnings("unchecked")
public abstract class BasePresenter<ListenerType extends BaseNavigation> extends Fragment {

    public interface ViewCreatedListener {
        void onViewCreated(View view);
    }

    protected final String LOG_TAG = getClass().getSimpleName();
    protected ViewCreatedListener viewCreatedListener;

    protected ListenerType callback;

    protected Context context;

    @Inject protected UseCaseHandler useCaseHandler;

    @Inject protected UseCaseProvider useCaseProvider;

    protected BaseView baseView;

    @Override
    public void onAttach(Context context) {
        try {
            callback = (ListenerType) getParentFragment();
            if (callback == null) {
                callback = (ListenerType) context;
            }
        } catch (ClassCastException e) {
            throw new RuntimeException(context + " must implement " + callback.getClass().getSimpleName());
        }

        this.context = context;
        if (context instanceof BaseActivity) {
            ((BaseActivity) this.context).component().inject((BasePresenter<BaseNavigation>) this);
        } else {
            throw new RuntimeException("baseActivity not instanceof BaseActivity");
        }

        super.onAttach(context);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        baseView = instanceView();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View mainView = inflater.inflate(baseView.getContentView(), container, false);
        setupPresenterLayout(mainView);
        ButterKnife.bind(this, mainView);
        baseView.setupLayout(mainView);
        return mainView;
    }

    private void setupPresenterLayout(View mainView) {

    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (viewCreatedListener != null) {
            viewCreatedListener.onViewCreated(view);
        }
        doOnViewCreated(view, savedInstanceState);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        loadViewData();
    }
    /**
     * @param viewCreatedListener the viewCreatedListener to set
     */
    public void setViewCreatedListener(ViewCreatedListener viewCreatedListener) {
        this.viewCreatedListener = viewCreatedListener;
    }

    protected void doOnViewCreated(View view, Bundle savedInstanceState) {

    }

    public boolean backPressed() {
        return false;
    }

    protected abstract BaseView instanceView();

    protected abstract void loadViewData();
}

File: app/src/main/java/es/santirivera/surveilfall/base/presenter/BasePresenter.kt
package es.santirivera.surveilfall.base.presenter

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import butterknife.ButterKnife
import es.santirivera.surveilfall.base.activity.BaseActivity
import es.santirivera.surveilfall.base.interfaces.BaseNavigation
import es.santirivera.surveilfall.base.view.BaseView
import es.santirivera.surveilfall.domain.usecases.base.UseCaseHandler
import es.santirivera.surveilfall.domain.usecases.providers.UseCaseProvider
import javax.inject.Inject


@Suppress("UNCHECKED_CAST")
abstract class BasePresenter<ListenerType : BaseNavigation> : Fragment() {

    private lateinit var callback: ListenerType

    @set:Inject
    lateinit var useCaseHandler: UseCaseHandler

    @set:Inject
    lateinit var useCaseProvider: UseCaseProvider

    private lateinit var baseView: BaseView<*>

    abstract val titleForActivity: String?


    override fun onAttach(context: Context) {
        super.onAttach(context)
        callback = try {
            if (parentFragment != null) {
                parentFragment as ListenerType
            } else {
                context as ListenerType
            }
        } catch (e: ClassCastException) {
            throw RuntimeException(context.toString() + " must implement " + callback.javaClass.simpleName)
        }
        if (context is BaseActivity) {
            (this.context as BaseActivity).component().inject(this as BasePresenter<BaseNavigation>)
        } else {
            throw RuntimeException("baseActivity not instanceof BaseActivity")
        }
    }

    override fun onResume() {
        super.onResume()
        val title = titleForActivity
        val showMenu = shouldShowMenu()
        val activity = activity as BaseActivity?
        if (title != null && title.isNotEmpty() && activity != null) {
            activity.title = title
        }
        activity!!.setDrawerEnabled(showMenu)
    }

    fun updateTitle() {
        val title = titleForActivity
        val activity = activity as BaseActivity?
        if (title != null && title.isNotEmpty() && activity != null) {
            activity.title = title
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        baseView = instanceView()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val mainView = inflater.inflate(baseView.contentView, container, false)
        ButterKnife.bind(this, mainView)
        baseView.setupLayout(mainView)
        return mainView
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        loadViewData()
    }


    fun backPressed(): Boolean {
        return false
    }

    protected abstract fun instanceView(): BaseView<*>

    protected abstract fun loadViewData()

    abstract fun shouldShowMenu(): Boolean
}
